storage_service/locations/models/dspace_rest.py

```diff
@@ -0,0 +1,379 @@
+
+        if package.package_type == 'DIP':
+            if 'dspace_dip_collection' in repository_collections:
+                dspace_collection = repository_collections['dspace_dip_collection']
+            else:
+                dspace_collection = self.dspace_dip_collection
+        elif package.package_type == 'AIP':
+            if 'dspace_aip_collection' in repository_collections:
+                dspace_collection = repository_collections['dspace_aip_collection']
+            else:
+                dspace_collection = self.dspace_aip_collection
+
+        if 'archivesspace_' + package.package_type.lower() + '_collection' in repository_collections:
+            archivesspace_collection = repository_collections['archivesspace_' + package.package_type.lower() +
+                                                              '_collection']
+        elif package.package_type == 'DIP':
+            archivesspace_collection = self.archivesspace_archival_object
+
+        LOGGER.info("Repo coll: %s", repository_collections)
+        item = {  # Structure necessary to create DSpace record
+            "type": "item",
+            "metadata": metadata
+        }
+
+        LOGGER.info("Item: %s", item)
+
+        # Headers for connecting to DSpace REST API
+        headers = {
+            'Accept': 'application/json',
+            "Content-Type": 'application/json',
+        }
+
+        collection_url = self.rest_url + '/collections/' + dspace_collection + '/items'
+
+        dspace_sessionid = self._login_to_dspace_rest()  # Logging in to REST api gives us a session id
+
+        # Create item in DSpace
+        try:
+            response = requests.post(collection_url,
+                                     headers=headers,
+                                     cookies={'JSESSIONID': dspace_sessionid},
+                                     data=json.dumps(item),
+                                     verify=False)
+        except Exception:
+            LOGGER.error('Could not create DSpace record: %s', collection_url)
+
+        dspace_item = response.json()
+
+        LOGGER.info('DSpace item: %s', dspace_item)
+
+        if package.package_type == 'DIP':
+            LOGGER.info('Package is DIP. Splitting package.')
+
+            for root, dirs, files in os.walk(source_path):
+                for name in files:
+                    with open(os.path.join(root, name), 'r') as f:
+                        content = f.read()
+
+                    bitstream_url = self.rest_url + '/items/' + dspace_item['uuid']
+                    bitstream_url += '/bitstreams?name=' + urllib.quote(name)
+
+                    try:
+                        requests.post(bitstream_url,
+                                      headers=headers,
+                                      cookies={'JSESSIONID': dspace_sessionid},
+                                      data=content,
```

Comment:
These are good points that you raise.

I'm not familiar with any way to do this by chunking.

I'll give this more thought and ask around.

Reply:
If I understand correctly, this might be a solution?:

 with open(os.path.join(root, name), 'rb') as content:
                            requests.post(bitstream_url,
                                      headers=headers,
                                      cookies={'JSESSIONID': dspace_sessionid},
                                      data=content,
                                      verify=False)